tests/runtime-tests/src/lib.rs

```diff
@@ -0,0 +1,400 @@
+                    "Test errored but not in the expected way.\n\texpected: {}\n\tgot: {}",
+                    expected,
+                    e
+                )
+            }
+        }
+        Response::Err(e) => {
+            error!(on_error, "Test errored: {e}");
+        }
+        Response::ErrNoBody => {
+            let stderr = spin.stderr.output_as_str().unwrap_or("<non-utf8>");
+            error!(on_error, "Spin did not return error body (most likely due to a test panicking). stderr:\n{stderr}");
+        }
+    }
+}
+
+/// Copies the test dir's manifest file into the temporary directory
+///
+/// Replaces template variables in the manifest file with components from the components path.
+fn copy_manifest(
+    test_dir: &Path,
+    components_path: &Path,
+    temp: &temp_dir::TempDir,
+) -> anyhow::Result<()> {
+    let manifest_path = test_dir.join("spin.toml");
+    let manifest = std::fs::read_to_string(&manifest_path).with_context(|| {
+        format!(
+            "no spin.toml manifest found in test directory {}",
+            test_dir.display()
+        )
+    })?;
+    let mut table = manifest
+        .parse::<toml::Table>()
+        .context("could not parse the manifest as toml")?;
+    for component in table["component"].as_array_mut().with_context(|| {
+        format!(
+            "malformed manifest '{}' does not contain 'component' array",
+            manifest_path.display(),
+        )
+    })? {
+        let source = component.get_mut("source").with_context(|| {
+            format!(
+                "malformed manifest '{}' does not contain 'source' string key in component",
+                manifest_path.display()
+            )
+        })?;
+        let source_str = source.as_str().with_context(|| {
+            format!(
+                "malformed manifest '{}' contains 'source' key that isn't a string in component",
+                manifest_path.display()
+            )
+        })?;
+        // TODO: make this more robust
+        if source_str.starts_with("{{") {
+            let name = &source_str[2..source_str.len() - 2];
+            let path = component_path(components_path, name);
+            let wasm_name = format!("{name}.wasm");
+            std::fs::copy(&path, temp.path().join(&wasm_name)).with_context(|| {
+                format!(
+                    "failed to copy wasm file '{}' to temporary directory",
+                    path.display()
+                )
+            })?;
+            *source = toml::Value::String(wasm_name);
+        }
+    }
+    let manifest =
+        toml::to_string(&table).context("failed to re-serialize manifest as a string")?;
+    std::fs::write(temp.path().join("spin.toml"), manifest)
+        .context("failed to copy spin.toml manifest to temporary directory")?;
+    Ok(())
+}
+
+#[derive(Debug)]
+enum Response {
+    Ok,
+    Err(String),
+    /// This happens when we panic
+    ErrNoBody,
+}
+
+impl FromStr for Response {
```

Comment:
I am curious why using raw tcp response manipulation instead of using some http library?